### PR TITLE
fixed problem with tuple obs being confused with return tuple

### PR DESCRIPTION
**Pull request by guillefix**
- [x] I have marked all applicable categories:
    + [x] exception-raising fix
    + [ ] algorithm implementation fix
    + [ ] documentation modification
    + [ ] new feature
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the code using `make commit-checks` (**required**)
- [ ] If applicable, I have mentioned the relevant/related issue(s)
- [ ] If applicable, I have listed every items in this Pull Request below

Envs with tuple observations failed when reset because the code thought that reset was returning a tuple of obs, return, done, info, so I added a better type check. I also made the continuous Critic work with structured obs (like tuples)

### CODE REVIEW

**Comment by Trinkle23897** [comment]
Have you read https://github.com/thu-ml/tianshou/pull/147#issuecomment-660956151?
Tuple obs space is not supported because of our design choice. It will lead to many undefined behaviors in `Batch`, and further slow down the entire pipeline (because of the memory layout of np.array with dtype=object). We suggest using either vectorized env wrapper or `collector.preprocess_fn` to convert tuple space to dict space.

**Comment by guillefix** [comment]
Ah, I hadn't seen that!
Oki, I will change to my environment to a Dict obs wrapped one.

I think the allowing the Critic to have structured obs (Dict) is still a good thing though no?

**Comment by Trinkle23897** [comment]
> I think the allowing the Critic to have structured obs (Dict) is still a good thing though no?

Yes, but we cannot assume which key the users want. Thus the network module is under `tianshou.utils.net` instead of `tianshou.net` to encourage the users to overwrite the forward function to fulfill their requirements.

So maybe you can change this PR to a documentation update to make things more clear for the incoming users?

**Comment by guillefix** [comment]
What I did is to have an option in the Critic network that just passes the obs as is to the preprocess_net, like the Actor network does as well, allowing the preprocess_net to be custom and deal with the Dict obs as desired by the user

**Comment by Trinkle23897** [comment]
I wrote something here: https://tianshou.readthedocs.io/en/master/tutorials/cheatsheet.html#user-defined-environment-and-different-state-representation

File: tianshou/utils/net/continuous.py [hunk 1]
```diff
@@ -122,18 +126,24 @@ def forward(
...
-                act,
+        if self.convert_to_torch:
+            obs = torch.as_tensor(
+                obs,
                 device=self.device,  # type: ignore
                 dtype=torch.float32,
-            ).flatten(1)
-            obs = torch.cat([obs, act], dim=1)
+            )
+            if self.flatten_obs: obs = obs.flatten(1)
```
Comment (Trinkle23897) [inline tianshou/utils/net/continuous.py:135]: If the `obs` from env is a dict, its type here is `Batch` and there's no `flatten` method.

**Comment by MischaPanch** [comment]
I think it's better to have as simple interfaces as possible. Gymnasium envs return `obs, info` on reset, and I don't think other signatures should be part of tianshou. Adding more input options to the source code (like Critic) also creates additional complexity, which can rather easily be delegated to user code (by preparing the env correctly). Can this PR be closed? @Trinkle23897 @guillefix

**Comment by MischaPanch** [comment]
Closing due to inactivity

**Pull request closed without merging**